extension/indexer/tx_indexer.go

```diff
@@ -0,0 +1,118 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package indexer
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/fees"
+)
+
+var (
+	_ AcceptedSubscriber = (*txIndexer)(nil)
+	_ AcceptedSubscriber = (*noopTxIndexer)(nil)
+)
+
+var (
+	failureByte = byte(0x0)
+	successByte = byte(0x1)
+)
+
+type TxIndexer interface {
+	Accepted(ctx context.Context, blk *chain.StatelessBlock) error
+	GetTransaction(txID ids.ID) (bool, int64, bool, fees.Dimensions, uint64, error)
+}
+
+type txIndexer struct {
+	db database.Database
+}
+
+func NewTxIndexer(db database.Database) *txIndexer {
+	return &txIndexer{db: db}
+}
+
+func (i *txIndexer) Accepted(_ context.Context, blk *chain.StatelessBlock) error {
+	batch := i.db.NewBatch()
+	defer batch.Reset()
+
+	timestamp := blk.GetTimestamp()
+	results := blk.Results()
+	for j, tx := range blk.Txs {
+		result := results[j]
+		if err := i.storeTransaction(
+			batch,
+			tx.ID(),
+			timestamp,
+			result.Success,
+			result.Units,
+			result.Fee,
+		); err != nil {
+			return err
+		}
+	}
+
+	return batch.Write()
+}
+
+func (*txIndexer) storeTransaction(
+	batch database.KeyValueWriter,
+	txID ids.ID,
+	t int64,
+	success bool,
+	units fees.Dimensions,
+	fee uint64,
+) error {
+	v := make([]byte, consts.Uint64Len+1+fees.DimensionsLen+consts.Uint64Len)
+	binary.BigEndian.PutUint64(v, uint64(t))
+	if success {
+		v[consts.Uint64Len] = successByte
+	} else {
+		v[consts.Uint64Len] = failureByte
+	}
+	copy(v[consts.Uint64Len+1:], units.Bytes())
+	binary.BigEndian.PutUint64(v[consts.Uint64Len+1+fees.DimensionsLen:], fee)
+	return batch.Put(txID[:], v)
+}
+
+func (i *txIndexer) GetTransaction(txID ids.ID) (bool, int64, bool, fees.Dimensions, uint64, error) {
+	v, err := i.db.Get(txID[:])
+	if errors.Is(err, database.ErrNotFound) {
+		return false, 0, false, fees.Dimensions{}, 0, nil
+	}
+	if err != nil {
+		return false, 0, false, fees.Dimensions{}, 0, err
+	}
+	t := int64(binary.BigEndian.Uint64(v))
+	success := true
+	if v[consts.Uint64Len] == failureByte {
+		success = false
+	}
+	d, err := fees.UnpackDimensions(v[consts.Uint64Len+1 : consts.Uint64Len+1+fees.DimensionsLen])
+	if err != nil {
+		return false, 0, false, fees.Dimensions{}, 0, err
+	}
+	fee := binary.BigEndian.Uint64(v[consts.Uint64Len+1+fees.DimensionsLen:])
+	return true, t, success, d, fee, nil
+}
+
+type noopTxIndexer struct{}
```

Comment:
nit: export type so caller can use this type if it wants to. Applicable to rest of concrete-types in this package

Reply:
I disagree with this. This is intentionally hiding that detail assuming callers choose to create this type and don't need to know the details or change anything about it after the fact.

Open to changing my mind on this though

Reply:
1. If you don't want to let the caller know about the type for some reason, then you can just return the interface. Returning an unexported struct makes the type awkward to use. You can't embed, wrap, return a non-exported type from a calling package.
2. At the same time what's the disadvantage of exporting the type? Returning a concrete type gives more flexibility to the caller to use it either as the interface or as the concrete type. I think we should let consumers of the package decide how they want to use our api (this way they can use our indexer interface, wrap it with their own, or use the type).

Reply:
Exported